### PR TITLE
Partially Revert "feat(mechanics): The accompany npc objective is no longer imp…

### PR DESCRIPTION
**Pull request by warp-core**
…licitly save (#7335)"

This reverts commit 7af610fd1416e6f90ec1e8bd890b76da14b0a1af.

**Bugfix:**

Thanks to @1010todd, @Terin, @Squiddy#8211, and @mOctave#5807 for helping find and investigate this bug.

## Fix Details
The commit being reverted here seems to have introduced a strange bug where, sometimes, upon loading a save with an active mission with an `npc evade` in the current system, `HasSucceeded()` would inappropriately return true (at least, this is my understanding of what was happening, I'm not entirely sure, though) resulting in the mission being completed without the player having to do anything but reload.

I have tested with the `Remnant: Defense 1` mission, but @Terin has also reported similar behaviour with Defend Sabik from the FW story.

## Testing Done
It doesn't seem to happen every time for me, but others have reported it more common.
Reload repeatedly without this PR and the mission will become completed fairly quickly.
Reload repeatedly with this PR and the expected behaviour of the mission not being completed occurs.

## Save File
I suggest the first file listed, but I've provided many for fun, I guess.:
[Oh Dear~3013-12-11-Remnantdefense1available.txt](https://github.com/endless-sky/endless-sky/files/9804391/Oh.Dear.3013-12-11-Remnantdefense1available.txt)

[Oh Dear.txt](https://github.com/endless-sky/endless-sky/files/9804383/Oh.Dear.txt)
[Oh Dear~~previous-1.txt](https://github.com/endless-sky/endless-sky/files/9804384/Oh.Dear.previous-1.txt)
[Oh Dear~~previous-2.txt](https://github.com/endless-sky/endless-sky/files/9804385/Oh.Dear.previous-2.txt)
[Oh Dear~~previous-3.txt](https://github.com/endless-sky/endless-sky/files/9804386/Oh.Dear.previous-3.txt)
[Oh Dear~3013-12-08-Remnantfirstcontactdone.txt](https://github.com/endless-sky/endless-sky/files/9804387/Oh.Dear.3013-12-08-Remnantfirstcontactdone.txt)
[Oh Dear~3013-12-11-6848351.txt](https://github.com/endless-sky/endless-sky/files/9804388/Oh.Dear.3013-12-11-6848351.txt)
[Oh Dear~3013-12-11-asdfasdfsda.txt](https://github.com/endless-sky/endless-sky/files/9804389/Oh.Dear.3013-12-11-asdfasdfsda.txt)
[Oh Dear~3013-12-11-Remnantdefense1acceptedconversationopen.txt](https://github.com/endless-sky/endless-sky/files/9804390/Oh.Dear.3013-12-11-Remnantdefense1acceptedconversationopen.txt)
[Oh Dear~3013-12-11-Remnantdefense1available0.9.16.txt](https://github.com/endless-sky/endless-sky/files/9804392/Oh.Dear.3013-12-11-Remnantdefense1available0.9.16.txt)
[Oh Dear~3013-12-11-Remnantdefense1launched.txt](https://github.com/endless-sky/endless-sky/files/9804393/Oh.Dear.3013-12-11-Remnantdefense1launched.txt)
[Oh Dear~3013-12-13-Remnantdefense1landedelsewhere.txt](https://github.com/endless-sky/endless-sky/files/9804396/Oh.Dear.3013-12-13-Remnantdefense1landedelsewhere.txt)
[Oh Dear~3013-12-15-Remnantdefense1returnedtoCaelian.txt](https://github.com/endless-sky/endless-sky/files/9804397/Oh.Dear.3013-12-15-Remnantdefense1returnedtoCaelian.txt)

### CODE REVIEW

File: source/NPC.cpp [hunk 1]
```diff
@@ -474,25 +477,23 @@ bool NPC::HasSucceeded(const System *playerSystem, bool ignoreIfDespawnable) con
...
+	// it cannot be disabled or destroyed, and must be in the player's system.
+	// Destroyed `accompany` are handled in HasFailed(). If the NPC block has
+	// `evade`, the ship can be disabled, destroyed, captured, or not present.
 	if(mustEvade || mustAccompany)
 		for(const shared_ptr<Ship> &ship : ships)
 		{
 			auto it = actions.find(ship.get());
-			// Captured or destroyed ships have either succeeded or no longer count.
-			if(it->second & (ShipEvent::DESTROY | ShipEvent::CAPTURE))
-				continue;
```
Comment (Amazinite) [inline source/NPC.cpp:487]: Try taking this if statement and nesting it in the if statement below. That is:
```c++
if(it != actions.end())
{
	// Captured or destroyed ships have either succeeded or no longer count.
	if(it->second & (ShipEvent::DESTROY | ShipEvent::CAPTURE))
		continue;
	// A ship that was disabled is considered 'immobile'.
	isImmobile = (it->second & ShipEvent::DISABLE);
	// If this NPC is 'derelict' and has no ASSIST on record, it is immobile.
	isImmobile |= ship->GetPersonality().IsDerelict()
		&& !(it->second & ShipEvent::ASSIST);
}
```

**Review by Amazinite — COMMENTED** [review]
If we are going to revert this then I'd say only revert the code since there's nothing wrong with the data changes.

File: data/hai/hai missions.txt [hunk 2]
```diff
@@ -1562,7 +1562,7 @@ mission "Pirate Troubles [4]"
 		"reputation: Hai Merchant (Human)" += 40
 		"reputation: Hai Merchant (Sympathizers)" += 40
 	
-	npc accompany save
+	npc save accompany
```
Comment (warp-core) [inline data/hai/hai missions.txt:1565]:
```suggestion
	npc accompany save
```

File: data/human/deep missions.txt [hunk 3]
```diff
@@ -2101,7 +2101,7 @@ mission "Deep: Scientist Rescue 1"
 			`	The communications device on his belt rings, and he quickly responds, giving orders to the fleet that will be joining you on the mission. Before you can return to your ship, he says, "Good luck, Captain, and may we all make it out of this alive. We'll meet you in orbit."`
 				accept
 	
-	npc accompany save
+	npc save accompany
```
Comment (warp-core) [inline data/human/deep missions.txt:2104]:
```suggestion
	npc accompany save
```

File: data/human/deep missions.txt [hunk 4]
```diff
@@ -2256,7 +2256,7 @@ mission "Deep: Scientist Rescue 2"
 		dialog "As your engines roar, you notice a large pirate ship looming over the planet. Even from the underside, it is undeniably a Bactrian, but how a pirate could ever attain one is a mystery."
 		dialog `	You receive a transmission from Lieutenant Paris. "I have a sneaking suspicion that Beelzebub may be behind this. I suggest we jump out of here immediately and make it back to the Deep alive before taking on that Bactrian."`
 	
-	npc accompany save
+	npc save accompany
```
Comment (warp-core) [inline data/human/deep missions.txt:2259]:
```suggestion
	npc accompany save
```

File: data/human/human missions.txt [hunk 5]
```diff
@@ -2922,7 +2922,7 @@ mission "Hauler VI cargo prototype 1"
 	destination "Skymoot"
 	to offer
 		"armament deterrence" > 0
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/human/human missions.txt:2925]:
```suggestion
	npc accompany save
```

File: data/human/human missions.txt [hunk 6]
```diff
@@ -2992,7 +2992,7 @@ mission "Hauler VI cargo prototype 2"
 	landing
 	to offer
 		has "Hauler VI cargo prototype 1: done"
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/human/human missions.txt:2995]:
```suggestion
	npc accompany save
```

File: data/human/human missions.txt [hunk 7]
```diff
@@ -3013,7 +3013,7 @@ mission "Hauler VI cargo prototype 3"
 	landing
 	to offer
 		has "Hauler VI cargo prototype 2: done"
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/human/human missions.txt:3016]:
```suggestion
	npc accompany save
```

File: data/remnant/remnant 2 cognizance.txt [hunk 8]
```diff
@@ -445,7 +445,7 @@ mission "Remnant: Cognizance 12"
 	source "Viminal"
 	destination "Esquiline"
 	stopover "Nasqueron"
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/remnant/remnant 2 cognizance.txt:448]:
```suggestion
	npc accompany save
```

File: data/remnant/remnant 2 cognizance.txt [hunk 9]
```diff
@@ -530,7 +530,7 @@ mission "Remnant: Cognizance 15"
 	destination "Viminal"
 	to offer
 		has "Remnant: Cognizance 14: active"
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/remnant/remnant 2 cognizance.txt:533]:
```suggestion
	npc accompany save
```

File: data/remnant/remnant 2 cognizance.txt [hunk 10]
```diff
@@ -553,7 +553,7 @@ mission "Remnant: Cognizance 16"
 			`	After several hours of watching the creature, one captain pushes his way forward to where you, Chilia, and Plume are talking. "I think I have an idea about this thing, Prefect!" the man signs excitedly. "Look at the positions of the glowing tips when it freezes. That pattern there matches the star positions of the main Ember Waste systems when viewed from Peragenor, and that second matches Convector." He pulls out a datapad and pulls up a projection of the starmap. "It is hard to tell due to the imprecision, but it looks like the third freeze is probably Antevorta or Nenia."`
 			`	Plume chimes in, "If those freezes indicate star positions, then the flying and landing behavior in between them may be indicating that it wants you to travel there. Please give it a try."`
 				accept
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/remnant/remnant 2 cognizance.txt:556]:
```suggestion
	npc accompany save
```

File: data/remnant/remnant 2 cognizance.txt [hunk 11]
```diff
@@ -617,7 +617,7 @@ mission "Remnant: Cognizance 18"
 			label letsgo
 			`	Chilia continues. "You don't need to actually land anywhere in Nenia this time, Captain. Just make sure the Pelicans get there, load up, and return safely." The two stand up and get ready to leave. "May the Embers burn bright for you, Captain." With that, they disembark and quickly clear the blast area around the landing pad. Around you a large group of Pelicans are warming up, each one synchronizing to your IFF and confirming their status.`
 				accept
-	npc accompany save
+	npc save accompany
```
Comment (warp-core) [inline data/remnant/remnant 2 cognizance.txt:620]:
```suggestion
	npc accompany save
```

File: data/remnant/remnant 2 side missions.txt [hunk 12]
```diff
@@ -550,7 +550,7 @@ mission "Remnant: Shattered Light 4"
 			`	The dust and debris settle within a few minutes, and soon you and the <ship> are skimming over the vegetation. The approach gives you ample time to appreciate just how massive the Shattered Light is; everything that one could need to lay the foundations for a new colony is packed into a single ship. You note large gashes in the hull, massive sensor arrays, and a shattered dome on top. Based on the mismatch of parts, this ship has been substantially modified from its original hull, and survived untold damage.`
 			`	As you slow to a hover alongside the ship, Ruach appears on your viewscreen. "Well met, Captain! Our theory worked! We managed to snap the Shattered Light back into the Primary Weft, and are now onboard. The systems are dubious, though, so we dare not shut anything down lest we cannot restart it. We would appreciate your assistance in escorting us back to Viminal."`
 				accept
-	npc accompany save
+	npc accompany
```
Comment (warp-core) [inline data/remnant/remnant 2 side missions.txt:553]:
```suggestion
	npc accompany save
```

**Comment by Hurleveur** [comment]
Can't we fix the issue instead of reverting the changes? I really wanted this new accompany behaviour, as well as Grey.

**Comment by warp-core** [comment]
I really think the desired behaviour should be added with a new objective.

Hai Reveal isn't merged yet, and even when it is, it still needs a lot of work, any changes necessary can be done alongside it.

**Comment by Hurleveur** [comment]
Okay, I'll just remove accompany from my mission in the UHai campaign then, because I absolutely dont want save with it.
Conceptually I really dont like this change because it means `accompany save` is the same as `accompany`, and, implicitly, I always believed accompany without save meant accompany *without save*.

**Comment by warp-core** [comment]
You can't really accompany something that no longer exists, can you?

Even if allowing the ships to die is something we keep, @tehhowch has said explicitly that it's not something that should be done in a bug fix release.

**Comment by warp-core** [comment]
> > Even if allowing the ships to die is something we keep, @tehhowch has said explicitly that it's not something that should be done in a bug fix release.
> 
> Totally agreed, but the release is already out, or at least I thought so?

Yes, and it's got a bug where `npc evade` doesn't work properly, and given that that bug is apparently as a result of accessing an out of bounds element of a map, we're probably lucky it hasn't caused full crashes yet. The plan seems to be to release a hotfix for that. Whether that be with this or #7407.

**Pull request merged**